checkoff

```diff
@@ -1,167 +1,162 @@
     lab = args.lab
-    _check_lab_id(lab)
 
-    data = _load_data()
-    users = data['users']
+    with db() as c:
+        c.execute('SELECT `student`, `facilitator`, `timestamp` FROM `checkoffs`'
+                  'WHERE `lab` = %s', lab)
 
-    for student in users:
-        if lab in users[student]:
-            print(student)
+    print('Checkoffs for lab {} ({})'.format(lab, data['labs'][lab]['fullname']))
 
+    for i in c:
+        print('{} checked off by {} on {}'.format(
+            i['student'], i['facilitator'], i['timestamp']
+        ))
 
-def _acquire_lock():
-    # Atomically check and create lockfile
-    # https://stackoverflow.com/a/33223732/5194666
-    try:
-        os.open(LOCK_FILE_PATH, os.O_CREAT | os.O_EXCL)
-        return True
-    except FileExistsError:
-        return False
 
+def student_completed_labs(args, data):
+    user = args.user
+
+    with db() as c:
+        c.execute('SELECT `lab`, `facilitator`, `timestamp` FROM `checkoffs`'
+                  'WHERE `student` = %s', user)
 
-def _release_lock():
-    os.remove(LOCK_FILE_PATH)
+    print('Checkoffs for student {} (track: {})'.format(user, data['roster'][user]))
 
+    for i in c:
+        print('Lab {} checked off by {} on {}'.format(
+            i['lab'], i['facilitator'], i['timestamp']
+        ))
 
-def _mail_checkoff(user, lab):
+
+def _mail_checkoff(user, lab, data):
     # subprocess needs input in bytes
-    msg = 'Congratulations {}, you have been checked off for lab {}'.format(user, lab).encode()
-    subject = 'Unix Sysadmin Decal Checkoff for Lab {}'.format(lab)
+    msg = 'Congratulations {}, you have been checked off for lab {} ({})'.format(
+        user, lab, data['labs'][lab]['fullname'],
+    ).encode()
+    subject = 'Unix SysAdmin DeCal Checkoff for Lab {}'.format(lab)
     recipient = '{}@ocf.berkeley.edu'.format(user)
     cc = 'decal+checkoffs@ocf.berkeley.edu'
-    subprocess.run(['mail', '-s', subject, '-c', cc, recipient], stdout=subprocess.PIPE, input=msg)
-
-
-def _load_data():
-    with open(DATA_PATH, 'r') as f:
-        data = json.load(f)
-    return data
-
-
-def _write_data(data):
-    with open(DATA_PATH, 'w') as f:
-        json.dump(data, f)
+    subprocess.run(['mail', '-s', subject, '-c', cc, recipient],
+                   stdout=subprocess.PIPE, input=msg)
 
 
-def _check_lab_id(lab):
-    if not lab.startswith('a') and not lab.startswith('b'):
-        raise ValueError('Lab "{}" not proper format (must start with a or b)'.format(lab))
+def main(argv=None):
+    data = _load_data()
 
+    def validate_lab(lab):
+        if lab in data['labs'].keys():
+            return lab
+        else:
+            raise argparse.ArgumentTypeError('Invalid lab: {}'.format(lab))
 
-def main(argv=None):
-    commands = {
-        'view': view,
-        'checkoff': checkoff,
-        'labs': list_labs,
-    }
+    def validate_user(user):
+        if user in data['roster'].keys() and user_exists(user):
```

Comment:
Same thing here, `keys()` is unnecessary.

Also, if there is a case where a user can be on the roster but doesn't have an OCF account, I think that warrants its own error msg. Will make debugging easier for whoever is running the script.

Reply:
students will have to make an OCF account for the course anyways, so this shouldn't be a problem. This code will basically only execute when people misspell usernames from their phoe keyboards anyways lol